File: examples/consumers/ruby-blog/webhouse.rb
# F125 reference reader for @webhouse/cms file-based content (Ruby).
#
# Reads JSON documents from content/{collection}/ and exposes them as plain
# Hash instances. Designed to be thin (stdlib only), and safe — slugs and
# collection names are validated against SAFE_NAME to prevent path traversal.
#
# Reference implementation for the future webhouse_cms RubyGems package.
# Drop into a Rails app under app/lib/webhouse.rb — the API is identical.

require 'json'
require 'pathname'

module Webhouse
  SAFE_NAME = /\A[a-z0-9]([a-z0-9-]*[a-z0-9])?\z/

  class InvalidName < StandardError; end

  class Reader
    attr_reader :content_dir

    def initialize(content_dir)
      @content_dir = Pathname.new(content_dir).expand_path
    end

    # List all published documents in a collection.
    # locale: optional locale filter; pass nil for all locales
    # Returns documents sorted by data.date descending.
    def collection(name, locale: nil)
      validate_name!(name)
      dir = @content_dir.join(name)
      return [] unless dir.directory?

      docs = dir.glob('*.json').filter_map do |f|
        doc = JSON.parse(f.read) rescue next
        next unless doc['status'] == 'published'
        next if locale && doc['locale'] != locale
        doc
      end

      docs.sort_by { |d| d.dig('data', 'date') || '' }.reverse
    end

    # Load a single published document by collection and slug.
    # Returns nil if not found, malformed, or unpublished.
    # Raises InvalidName on path traversal attempts.
    def document(collection_name, slug)
      validate_name!(collection_name)
      validate_name!(slug)

      file = @content_dir.join(collection_name, "#{slug}.json").expand_path
      raise InvalidName, "path escapes content dir" unless file.to_s.start_with?(@content_dir.to_s)
      return nil unless file.file?

      doc = JSON.parse(file.read)
      doc['status'] == 'published' ? doc : nil
    rescue JSON::ParserError
      nil
    end

    # Find the sibling translation of a document via translationGroup.
    def find_translation(doc, collection_name)
      tg = doc['translationGroup']
      return nil unless tg
      collection(collection_name).find do |other|
        other['translationGroup'] == tg && other['locale'] != doc['locale']
      end
    end

    # Convenience: returns the singleton globals/site.json doc, cached.
    def globals
      @globals ||= document('globals', 'site') || {}
    end

    private

    def validate_name!(name)
      raise InvalidName, "invalid name '#{name}'" unless name.is_a?(String) && name.match?(SAFE_NAME)
    end
  end

  # Helpers for templates
  def self.string(doc, key, default = '')
    return default unless doc.is_a?(Hash)
    value = doc.dig('data', key)
    value.is_a?(String) ? value : default
  end

  def self.tags(doc)
    return [] unless doc.is_a?(Hash)
    value = doc.dig('data', 'tags')
    value.is_a?(Array) ? value.select { |v| v.is_a?(String) } : []
  end
end

File: examples/consumers/elixir-blog/lib/webhouse_blog/webhouse.ex
defmodule WebhouseBlog.Webhouse do
  @moduledoc """
  F125 reference reader for @webhouse/cms file-based content (Elixir).

  Reads JSON documents from `content/{collection}/` directories. Designed to
  be thin (Jason for JSON parsing) and safe — slugs and collection names are
  validated against `@safe_name` to prevent path traversal.

  Drop into a Phoenix controller and the API is identical — Plug and Phoenix
  share the same Conn pipeline.
  """

  @safe_name ~r/^[a-z0-9]([a-z0-9-]*[a-z0-9])?$/

  defmodule InvalidName do
    defexception [:name]
    def message(%{name: name}), do: "invalid name '#{name}'"
  end

  def content_dir do
    Path.join([File.cwd!(), "content"])
  end

  defp validate!(name) when is_binary(name) do
    unless Regex.match?(@safe_name, name) do
      raise InvalidName, name: name
    end
  end

  defp validate!(_), do: raise(InvalidName, name: "non-string")

  @doc "List published documents in a collection. Pass locale as nil for all."
  def collection(name, locale \\ nil) do
    validate!(name)
    dir = Path.join(content_dir(), name)

    case File.ls(dir) do
      {:ok, entries} ->
        entries
        |> Enum.filter(&String.ends_with?(&1, ".json"))
        |> Enum.map(fn file ->
          case File.read(Path.join(dir, file)) do
            {:ok, raw} ->
              case Jason.decode(raw) do
                {:ok, doc} -> doc
                _ -> nil
              end

            _ ->
              nil
          end
        end)
        |> Enum.reject(&is_nil/1)
        |> Enum.filter(&(Map.get(&1, "status") == "published"))
        |> Enum.filter(fn doc ->
          locale == nil or Map.get(doc, "locale") == locale
        end)
        |> Enum.sort_by(fn doc -> get_in(doc, ["data", "date"]) || "" end, :desc)

      _ ->
        []
    end
  end

  @doc "Load a single published document by collection + slug, or nil."
  def document(collection, slug) do
    validate!(collection)
    validate!(slug)

    path = Path.join([content_dir(), collection, "#{slug}.json"])
    abs = Path.expand(path)

    if not String.starts_with?(abs, content_dir()) do
      nil
    else
      case File.read(abs) do
        {:ok, raw} ->
          case Jason.decode(raw) do
            {:ok, doc} ->
              if Map.get(doc, "status") == "published", do: doc, else: nil

            _ ->
              nil
          end

        _ ->
          nil
      end
    end
  end

  @doc "Find the sibling translation of a document via translationGroup."
  def find_translation(doc, collection) do
    case Map.get(doc, "translationGroup") do
      nil ->
        nil

      tg ->
        Enum.find(collection(collection), fn other ->
          Map.get(other, "translationGroup") == tg and
            Map.get(other, "locale") != Map.get(doc, "locale")
        end)
    end
  end

  @doc "Cached singleton globals/site.json document. Loaded once per request."
  def globals do
    document("globals", "site") || %{"data" => %{}}
  end

  @doc "Safely extract a string field from doc.data."
  def string(doc, key, default \\ "")

  def string(nil, _, default), do: default

  def string(doc, key, default) do
    case get_in(doc, ["data", key]) do
      v when is_binary(v) -> v
      _ -> default
    end
  end
end
